feat: adding currency value component

jpuri opened the pull request:
## **Description**
Adding value component for currency.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23260?quickstart=1)

## **Related issues**

Fixes: MetaMask/metamask-extension#23157

## **Manual testing steps**
NA

## **Screenshots/Recordings**
![Screenshot 2024-03-01 at 11 34 23 AM](https://github.com/MetaMask/metamask-extension/assets/2182307/fdc41201-7e3b-4516-9257-f142bb269bf2)

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

## **Pre-merge author checklist**

- [X] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've clearly explained what problem this PR is solving and how it is solved.
- [X] I've linked related issues
- [X] I've included manual testing steps
- [X] I've included screenshots/recordings if applicable
- [X] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [X] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [X] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [X] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

jpuri committed:
Adding currency value component

jpuri committed:
Update

jpuri committed:
Merge branch 'develop' into currency_value_component

jpuri committed:
Update

jpuri committed:
Merge branch 'currency_value_component' of github.com:MetaMask/metamask-extension into currency_value_component

jpuri committed:
Merge branch 'develop' into currency_value_component

jpuri committed:
Merge branch 'develop' into currency_value_component

cryptotavares reviewed (changes requested):
Due to a typo `numberOfDecimals` is not being taken into account on `CurrencyDisplay`. Plus we can have TS proper type checking react propTypes from JS components and stop casting them to `any`

jpuri commented:
Hey @cryptotavares : about `numberOfDecimals` I have kept configuration params to a minimum for now - adding only those that we need on confirmation pages.

If we do not pass number of decimals it uses default from user setting or from currency settings.

jpuri committed:
Update

cryptotavares commented on ui/components/ui/currency-display/currency-display.component.js:
```diff
@@ -11,6 +11,7 @@ import {
   TextVariant,
 } from '../../../helpers/constants/design-system';
 
+/** @param {PropTypes.InferProps<typeof propTypes>>} */
```
```suggestion
/** @param {PropTypes.InferProps<typeof CurrencyDisplayPropTypes>>} */
```

jpuri replied:
I made the change.

cryptotavares committed:
fix: infer types via propTypes

Remove CurrencyDisplay cast to any
and infer types via propTypes.
Add to disable two jsdoc linting rules
but the added benefit far outweighs
disabling those rules.

jpuri committed:
Update

jpuri committed:
Update

jpuri commented on ui/components/app/confirm/info/row/currency.tsx:
```diff
@@ -0,0 +1,44 @@
+
+// Below is to avoid TS issues in using these components
+const UserCurrDisplay = UserPreferencedCurrencyDisplay as any;
+const CurrDisplay = CurrencyDisplay as any;
```
Changes are committed.

jpuri committed:
Update

jpuri committed:
Merge branch 'develop' into currency_value_component

cryptotavares commented on ui/components/app/user-preferenced-currency-display/user-preferenced-currency-display.component.js:
```diff
@@ -86,3 +90,6 @@ UserPreferencedCurrencyDisplay.propTypes = {
    */
   ...CurrencyDisplay.propTypes,
```
This seems to be breaking the type inference... I have noticed that for some reason using `Object.assign({}, UserPreferencedCurrencyDisplayPropTypes, CurrencyDisplay)` works... but ends up breaking linting for functional component opts destruct.
Probably the best way to deal with this is to be explicit about the propTypes, instead of extending from a different component propTypes. For example `UserPreferencedCurrencyDisplayPropTypes` would look like this:
```js
const UserPreferencedCurrencyDisplayPropTypes = {
  className: PropTypes.string,
  'data-testid': PropTypes.string,
  prefix: PropTypes.string.isRequired,
  value: PropTypes.string,
  numberOfDecimals: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  hideLabel: PropTypes.bool,
  hideTitle: PropTypes.bool,
  style: PropTypes.object,
  showEthLogo: PropTypes.bool,
  ethLogoHeight: PropTypes.number,
  type: PropTypes.oneOf([PRIMARY, SECONDARY]),
  ethNumberOfDecimals: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  fiatNumberOfDecimals: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  showFiat: PropTypes.bool,
  showNative: PropTypes.bool,
  showCurrencySuffix: PropTypes.bool,
  currency: PropTypes.string,
  denomination: PropTypes.oneOf([
    EtherDenomination.GWEI,
    EtherDenomination.ETH,
  ]),
  displayValue: PropTypes.string,
  prefixComponent: PropTypes.node,
  suffix: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  prefixComponentWrapperProps: PropTypes.object,
  textProps: PropTypes.object,
  suffixProps: PropTypes.object,
};
```

segun replied:
This pattern was introduced by the another PR. I propose we fix this in another PR (unless @jpuri is fine with fixing it here).

cryptotavares replied:
@segun although it was introduced in a different PR, it is this PR that breaks type inference. And because of that, it should be fixed in this PR.

It breaks type inference, because on this PR we have introduced `PropTypes.InferProps<typeof propTypes>>` as way for TS to properly infer the functional component props type from propTypes (see [here](https://github.com/MetaMask/metamask-extension/pull/23260/files#diff-c99d34b0461ac441cee8391fdf35e9622ab2a18c19820f748a2536a89f91d2c7R13)). Allowing us to benefit from TS without having to convert all jsx into tsx.

jpuri replied:
Good point I made the update.

segun commented on ui/components/app/confirm/info/row/currency.tsx:
```diff
@@ -0,0 +1,43 @@
+    {currency ? (
+      <CurrencyDisplay
+        currency={currency}
+        value={value ? `${value}` : undefined}
```
is there a use case for an undefined value? if yes, since it's currency should we display 0.00 instead?

cryptotavares replied:
Oh this was a good point. `value` is mandatory and typed as [`string | number`](https://github.com/MetaMask/metamask-extension/pull/23260/files#diff-f5b2f432afcf046ad259498c46bc5c67bac8fc75964c7d5b3d80c2eeea623b2cR13), so why are we checking if it is defined?

Can't we just:
```js
value={`${value}`}
```

jpuri replied:
I updated value to `value={`${value}`}`

jpuri committed:
Update

jpuri committed:
Merge branch 'currency_value_component' of github.com:MetaMask/metamask-extension into currency_value_component

jpuri committed:
Update

cryptotavares reviewed (approved):
🚀 🟢
